BlobContainer: add copyBlob method

bcully opened the pull request:
If a container implements copyBlob, then the copy is performed by the store, without client-side IO. If the store does not provide a copy operation then the default implementation throws UnsupportedOperationException.

This change provides implementations for the FS and S3 blob containers. More will follow.

Refs ES-11059

bcully committed:
BlobContainer: add copyBlob method

If a container implements copyBlob, then the copy is
performed by the store, without client-side IO. If the store
does not provide a copy operation then the default implementation
throws UnsupportedOperationException.

This change provides implementations for the FS and S3 blob containers.
More will follow.

elasticsearchmachine commented:
Pinging @elastic/es-distributed-indexing (Team:Distributed Indexing)

bcully commented:
run elasticsearch-ci/third-party / s3

bcully commented:
@elasticmachine test this please

bcully commented:
@elasticmachine run elasticsearch-ci/third-party / s3

bcully commented:
@elasticmachine test this please

bcully committed:
trigger s3 tests (revert before merge)

Commit:
[CI] Auto commit changes from spotless

bcully commented on .buildkite/pipelines/pull-request/s3.yml:
```diff
@@ -0,0 +1,14 @@
+steps:
+  - label: third-party / s3
+    command: |
+      export amazon_s3_bucket=elasticsearch-ci.us-west-2
+      export amazon_s3_base_path=$BUILDKITE_BRANCH
+      .ci/scripts/run-gradle.sh s3ThirdPartyTest
+    env:
+      USE_3RD_PARTY_S3_CREDENTIALS: "true"
+    timeout_in_minutes: 30
+    agents:
+      provider: gcp
+      image: family/elasticsearch-ubuntu-2004
+      machineType: n2-standard-8
+      buildDirectory: /dev/shm/bk
```
I added this in to trigger tests on S3. I'll revert it before merging anything.

bcully commented:
s3 test passed on real s3: https://gradle-enterprise.elastic.co/s/aymzd6ke25njw/tests/task/:modules:repository-s3:s3ThirdPartyTest/details/org.elasticsearch.repositories.s3.S3RepositoryThirdPartyTests/testCopy?top-execution=1

bcully committed:
FsBlobContainerTests: small improvements to testCopy

DaveCTurner reviewed (dismissed):
Please could you also extend the repository analyzer (`RepositoryAnalyzeAction` and friends) to cover this API?

DaveCTurner commented:
> Please could you also extend the repository analyzer (`RepositoryAnalyzeAction` and friends) to cover this API?

Sorry, meant to add more detail before submitting that review:

---

I'd suggest doing something with `BlobAnalyzeAction` to sometimes copy the blob it just wrote and then perform (some of?) the reads from the copy. Ideally we'd be checking for atomicity of reads from the copy just like we do today, making sure that we never see a partiallly-copied blob.

Skipping checks on a UOE is ok, there's prior art for that:

https://github.com/elastic/elasticsearch/blob/d2e667004596544bbbb594c7d42e215483caf9c5/x-pack/plugin/snapshot-repo-test-kit/src/main/java/org/elasticsearch/repositories/blobstore/testkit/analyze/UncontendedRegisterAnalyzeAction.java#L127-L132

bcully committed:
debugging FsBlobContainerTests::testCopy on ci

bcully committed:
Merge branch 'main' of github.com:elastic/elasticsearch into ES-11059

bcully committed:
is it sourceContainer vs targetContainer?

Tim-Brooks commented on modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java:
```diff
@@ -317,6 +318,57 @@ public void writeBlobAtomic(OperationPurpose purpose, String blobName, BytesRefe
+            throw new IllegalArgumentException("target blob container must be a S3BlobContainer");
+        }
+        if (failIfAlreadyExists) {
+            throw new UnsupportedOperationException("S3 blob container does not support failIfAlreadyExists");
```
This is probably outside of the scope of the ticket, but I'm pretty sure S3 does now support this. At least for Puts. https://aws.amazon.com/about-aws/whats-new/2024/08/amazon-s3-conditional-writes/

bcully replied:
As far as I could tell, while this is true for puts, it is not for copy. On put, there is
[If-None-Match](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html#API_PutObject_RequestSyntax)

    Uploads the object only if the object key name does not already exist in the bucket specified. Otherwise, Amazon S3 returns a 412 Precondition Failed error.`

But on copy, there only appears to be a set of [preconditions on the source](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CopyObject.html), not the target.

DaveCTurner replied:
I would rather we didn't add this argument if it's not going to be supported on S3. Instead, the caller needs to find a way to use this feature that does not require it to fail if the blob already exists. Typically that means we pick a unique fresh name for the blob after we've decided on its contents, such that if the blob _does_ already happen to exist then the operation already succeeded, the blob contents are already correct, and thus retrying the operation will do no harm.

I know we have had this flag on the `writeBlob` API for years but this was IMO a mistake given the lack of support for this feature in S3 at the time. We don't actually care any more if the blob exists, because of our move towards unique fresh names in all other places.

bcully replied:
I don't mind removing it. The specific workload motivating this PR is to be able to seed a new shard from the contents of an old one when splitting, and for that the source names have the uniqueness we want, and it is probably even convenient for retry to not fail if we happen to copy the same source file to the same destination twice.

It would be nice for safety to be able to assert that if the file is already present its etag matches what we expect and fail otherwise, since that indicates a logic error and potential data loss. I don't think that we can do that atomically with S3 right now, and I'm on the fence about whether it's worth doing it non-atomically (heading the destination before starting the copy). On the one hand it doubles the request count for what's expected to be a hit rate of zero if we're not buggy, but on the other hand split is going to be a relatively infrequent operation so the total request count may not really matter much.

DaveCTurner replied:
I'd suggest not doing this belt-and-braces approach, at least not in the production code. We should be able to catch this sort of problem in tests with a sufficiently picky test fixture instead. I would expect this kind of bug would eventually manifest as a corruption in Lucene itself if it really did happen in production.

bcully replied:
I've removed `failIfAlreadyExists` in ec271f5

Tim-Brooks commented on modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java:
```diff
@@ -317,6 +318,57 @@ public void writeBlobAtomic(OperationPurpose purpose, String blobName, BytesRefe
+    public void copyBlob(
+        OperationPurpose purpose,
+        String sourceBlobName,
+        BlobContainer targetBlobContainer,
```
Should we name this destination? The feels like a more common terminology for copying.

bcully replied:
sure, I can rename these.

bcully replied:
Done in 89ae18a

Tim-Brooks commented on modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java:
```diff
@@ -317,6 +318,57 @@ public void writeBlobAtomic(OperationPurpose purpose, String blobName, BytesRefe
+     *
+     * Server-side copy can be done for any size object, but if the object is larger than 5 GB then
+     * it must be done through a series of part copy operations rather than a single blob copy.
+     * See <a href="https://docs.aws.amazon.com/AmazonS3/latest/API/API_CopyObject.html">CopyObject</a>.
```
How does this fail exactly? Is the plan for this to be handled in a follow-up (supporting multi part copy)? Or are users of this supposed to catch this type of error and then resort to GET + PUT.

DaveCTurner replied:
I don't think we want our long-term plan to involve throwing a `UnsupportedOperationException` in this case. We'll almost certainly forget that we need to handle a UOE in the caller at some point, and given the rarity (but not impossibility) of encountering blobs that exceed 5GiB we will likely first notice when it causes a production outage. It's unreasonable to expect unit tests to cover such large cases. I see that a UOE is ok for now but we need to make this method work either always or never in the long run.

Please could we add a configurable limit on the maximum `CopyObject` size so we don't rely on the S3 API returning an error, and so that we can reduce the limit to something tiny in tests?

bcully replied:
FWIW I don't think we would throw UOE if the source object were over 5GB. That's an S3-side limit and I'd expect that to come back as a regular failed operation (a subclass of IOException IIRC).

I will test manually on S3 to get the concrete error it produces for larger than 5GB files and make test coverage of that case by generating the same error in a fixture. I'll also just go ahead and add multipart copy support now, triggered by receiving that error from S3. Later we can add a variant of `copyBlob` that takes a source size if known to avoid having to make a first copy attempt and handle failure, if it seems worthwhile.

DaveCTurner replied:
Ah yeah sorry I got the two threads mixed up. No UOE here indeed, just some kind of `AmazonS3Exception` or something. But still the point about this being a rarely-hit and easily-forgotten path remains, so yes please it'd be ideal if we could do multipart copy at the same time.

bcully replied:
My own notes here from manual testing: attempting to copy an object larger than 5g with `aws s3api copy-object --debug` produces an HTTP 200 `InvalidRequest` error response:
```
Response headers: {'x-amz-request-id': 'CZWYF5KNMSB99PP3', 'x-amz-id-2': 'vmTYh6g26hF/wU5PspZopMGFOepUe0tIzmU/9QT4DAe2BlvOW1Qk1raP5e46zTdsRImZZhyzB4tUtrAYPflUrQ==', 'Content-Type': 'application/xml', 'Transfer-Encoding': 'chunked', 'Date': 'Sat, 29 Mar 2025 00:06:49 GMT', 'Connection': 'close', 'Server': 'AmazonS3'}
Response body:
b'<Error><Code>InvalidRequest</Code><Message>The specified copy source is larger than the maximum allowable size for a copy source: 5368709120</Message><RequestId>CZWYF5KNMSB99PP3</RequestId><HostId>vmTYh6g26hF/wU5PspZopMGFOepUe0tIzmU/9QT4DAe2BlvOW1Qk1raP5e46zTdsRImZZhyzB4tUtrAYPflUrQ==</HostId></Error>'
```

bcully replied:
I've added multipart copy, adjusted the copyBlob signature to take a length, and branched to multipart copy when the length is too long for plain copyObject in [f4aad52](https://github.com/elastic/elasticsearch/pull/125737/commits/f4aad5251f53b626b7bb33186c0882a8ad5c672c).

Tim-Brooks commented on test/framework/src/main/java/org/elasticsearch/repositories/blobstore/ESMockAPIBasedRepositoryIntegTestCase.java:
```diff
@@ -360,6 +358,13 @@ public HttpStatsCollectorHandler(HttpHandler delegate) {
+        public HttpStatsCollectorHandler(HttpHandler delegate, String[] operations) {
+            this.delegate = delegate;
+            for (String operation : operations) {
+                operationCount.put(operation, 0L);
```
What is this change? Just pre-initializing to 0? If we preinitialize here, when are we then still doing `operationCount.getOrDefault(requestType, 0L) + 1`?

bcully replied:
yes, I noticed that we are currently filtering out 0 stats from the actual metrics because we only lazily create them on the expected metrics, and it seemed more symmetric with the actual metrics to prepopulate the expected ones. You're right that we shouldn't need `getOrDefault` any more, I'll replace that.

bcully replied:
Done in 27656b2

bcully replied:
And faa0a83

DaveCTurner replied:
Would you pull these out into a separate PR @bcully? Happy to review it (or even pull it out myself if you'd prefer), just seems nicer to keep this one focussed on the `copyBlob` implementation.

bcully replied:
I should have said how I noticed that we're filtering out 0 stats (removed in the hunk just above this). It's that without this change, things like `S3BlobStoreRepositoryTests::testRequestStats` would break:
```
Expected :{HeadObject=3, PutMultipartObject=36, GetObject=120, ListObjects=12, PutObject=114, DeleteObjects=6}
Actual   :{CopyObject=0, PutMultipartObject=36, HeadObject=3, PutObject=114, DeleteObjects=6, ListObjects=12, GetObject=120}
```
The fixes I saw were to either add additional filtering or preinitialize the metrics (generating non-zero copy metrics for this test didn't seem like an option since this test appears to be running higher level API operations and none of those use copy yet), and it seemed like preinitializing was cleaner and only a little longer.

I'm happy to replace this with additional filtering if you prefer it, just wanted to provide some context I should have noted earlier.

bcully committed:
Use get instead of getOrDefault for metrics in repo integ test

We are now preinitializing the metric map rather than updating it
lazily.

bcully committed:
copyBlob: use "destination" instead of "target"

bcully committed:
Merge branch 'main' of github.com:elastic/elasticsearch into ES-11059

bcully committed:
Remove failIfAlreadyExists from BlobContainer.copyBlob

bcully committed:
Add copyBlob to ESBlobStoreRepositoryIntegTestCase

bcully committed:
Preinitialize Azure metrics for BlobStoreRepositoryTests too

And remove the constructor that depends on lazy initialization.

bcully committed:
Merge branch 'main' of github.com:elastic/elasticsearch into ES-11059

bcully committed:
Add support for S3 multipart copy

bcully committed:
Change copyBlob API to be explicitly copyFrom

s3's multipart copy initiates the multipart operation on the
destination, so this avoids confusion.

Also include the blob length in the API to make it convenient for S3.
We can add a no-length API later if we don't always have it at hand.

bcully committed:
Use multipart copy if source blob size is too large for copyObject

Commit:
[CI] Auto commit changes from spotless

bcully committed:
spotless

bcully committed:
Merge branch 'ES-11059' of github.com:bcully/elasticsearch into ES-11059

bcully committed:
Update GoogleCloudStorageStatsCollectorHttpHandler constructor

bcully commented:
testCopy/testMultipartCopy passed on real s3: https://gradle-enterprise.elastic.co/s/ywuiqiqcaf5gw/tests/overview

bcully committed:
Merge branch 'main' of github.com:elastic/elasticsearch into ES-11059

bcully committed:
FSBlobContainerTests: workaround for occasional testCopy failure on CI

bcully committed:
Merge branch 'main' of github.com:elastic/elasticsearch into ES-11059

DaveCTurner commented:
I would rather it happened at the same time if at all possible. We have a bit of a habit of not quite finding the time to do this kind of follow-up later on, and gaps in the repo analysis process are a massive headache in the long run.

Repo analysis doesn't have particularly strong bwc guarantees so don't worry too much about getting a perfect design up front. As long as we hit the `CopyObject` API I'll be happy - I'm less concerned about stringent atomicity checks or anything like that.

bcully commented:
> I would rather it happened at the same time if at all possible. We have a bit of a habit of not quite finding the time to do this kind of follow-up later on, and gaps in the repo analysis process are a massive headache in the long run.
> 
> Repo analysis doesn't have particularly strong bwc guarantees so don't worry too much about getting a perfect design up front. As long as we hit the `CopyObject` API I'll be happy - I'm less concerned about stringent atomicity checks or anything like that.

Alright, I'll push something today. It may not have coverage of overwriting the source during the copy but the skeleton should be in place.

bcully committed:
BlobAnalyzeAction: rough first cut at copy

The control flow is a bit of a mess.

bcully commented:
I'm frankly a bit twisted around by the control flow in `BlobAnalyzeAction` but I've pushed something anyway just as a basis for conversation.

bcully committed:
GoogleCloudStorageBlobStoreRepositoryTests: post-merge fixup

bcully committed:
S3 copyBlob: convert amazon exception to NoSuchFileException when source is not found

Commit:
[CI] Auto commit changes from spotless

DaveCTurner commented on x-pack/plugin/snapshot-repo-test-kit/src/main/java/org/elasticsearch/repositories/blobstore/testkit/analyze/RepositoryAnalyzeAction.java:
```diff
@@ -521,6 +521,13 @@ public void run() {
                 final long targetLength = blobSizes.get(i);
                 final boolean smallBlob = targetLength <= MAX_ATOMIC_WRITE_SIZE; // avoid the atomic API for larger blobs
                 final boolean abortWrite = smallBlob && request.isAbortWritePermitted() && rarely(random);
+                final boolean doCopy = random.nextBoolean();
```
I think I'd be inclined to have the `RepositoryAnalyzeAction` control the name of the copied blob rather than just setting a flag. That way we can record it in `AsyncAction#expectedBlobs` and check it exists in the listing at the end.

Also this can be a rare action, we don't need to copy half of the blobs (+50% storage size on average).

DaveCTurner replied:
Actually on reflection it's probably simplest not to put this in `expectedBlobs` (for now anyway). Not only do we have to think about the `doEarlyCopy` case but also we need to handle the fact that GCS/Azure will fail the copy with a UOE.

I'd still be inclined to have the `RepositoryAnalyzeAction` specify the blob name rather than just set a flag in the request because we'll want to do these things in future, but that's the only change I'd suggest in this areal.

DaveCTurner replied:
I pushed d9db70da06563a10d6753843445c86c279a87a31 which should fix the `FailureIT`

DaveCTurner commented on x-pack/plugin/snapshot-repo-test-kit/src/main/java/org/elasticsearch/repositories/blobstore/testkit/analyze/BlobAnalyzeAction.java:
```diff
@@ -230,6 +258,7 @@ private static class BlobAnalysis {
                 checksumStart = randomLongBetween(0L, request.targetLength);
                 checksumEnd = randomLongBetween(checksumStart + 1, request.targetLength + 1);
             }
+            doEarlyCopy = random.nextBoolean();
```
Nice, but consider this an optional bonus feature, I'd be happy even if we only covered the copy-after-write path. Note that we can only track the copy as an expected blob in `RepositoryAnalyzeAction` if it doesn't potentially fail because of an early copy, so we'd need to pull this parameter up to the top level too.

bcully replied:
Would you prefer that I remove early copy from this PR then?

DaveCTurner commented on test/fixtures/s3-fixture/src/main/java/fixture/s3/S3HttpHandler.java:
```diff
@@ -201,10 +201,24 @@ public void handle(final HttpExchange exchange) throws IOException {
-                blobs.put(request.path(), blob.v2());
-                exchange.getResponseHeaders().add("ETag", blob.v1());
-                exchange.sendResponseHeaders(RestStatus.OK.getStatus(), -1);
+                // a copy request is a put request with a copy source header
```
Do we not have to do anything special here for the multipart copy implementation?

bcully replied:
hmm yes, I must be missing test coverage here.

DaveCTurner replied:
I would expect `org.elasticsearch.repositories.blobstore.testkit.analyze.S3RepositoryAnalysisRestIT` to cover this, but not every time

bcully replied:
I added a test override to drop the multipart copy threshold down to 1MB (symmetric with the multipart upload override). On my laptop I found that was still too large to hit it reliably so I tested it locally at 10K, at which point testWriteMaybeCopyRead would fail about 10% of the time with the original code here. It passes with https://github.com/elastic/elasticsearch/pull/125737/commits/cb2e324951095cfee60ddd8863d3d325cf35b799 at the 10k threshold (which is technically illegal - S3 documents a minimum of 5MB - but I think for testing it's fine).

DaveCTurner reviewed (commented):
I like it. Left a couple of small comments.

DaveCTurner committed:
Fix up BWC

DaveCTurner committed:
Fix failure-path test

bcully committed:
PR feedback: pass copy blob name as request parameter.

Commit:
[CI] Auto commit changes from spotless

bcully committed:
Exercise multipart copy randomly in testWriteMaybeCopyRead

bcully committed:
Merge commit '728eb75' into ES-11059

bcully committed:
Revert "trigger s3 tests (revert before merge)"

This reverts commit 9c9eeb30299061780dee375d103b8e797f4bd315.

DaveCTurner commented on test/fixtures/s3-fixture/src/main/java/fixture/s3/S3HttpHandler.java:
```diff
@@ -467,6 +508,27 @@ static List<String> extractPartEtags(BytesReference completeMultipartUploadBody)
 
+    private static final Pattern rangePattern = Pattern.compile("^bytes=([0-9]+)-([0-9]+)$");
+
+    private static Tuple<Long, Long> parsePartRange(final HttpExchange exchange) {
```
Could we use `org.elasticsearch.test.fixture.HttpHeaderParser#parseRangeHeader` here? Although I do like how this pattern is anchored at start and end, maybe we should do that in `HttpHeaderParser` too.

bcully replied:
Ah sure, I hadn't seen it, will do. I am inclined not to change that pattern here but happy to do it as a self-contained change.

DaveCTurner commented on modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java:
```diff
@@ -300,6 +304,11 @@ long getLargeBlobThresholdInBytes() {
     }
 
+    // package private for testing
+    long getLargeCopyThresholdInBytes() {
+        return MAX_FILE_SIZE.getBytes();
+    }
```
Could we make this a setting similar to the `buffer_size` and `chunk_size` ones? Ideally imposing the same limitations as S3 proper (so a minimum of 5MiB and a max of 5GiB)? That way we can adjust it in `S3RepositoryAnalysisRestIT` to get the coverage of both paths there.

bcully replied:
Done in https://github.com/elastic/elasticsearch/pull/125737/commits/a9ff4b926ffd68d060271fbb6967b2d108b7edab

bcully committed:
Use HttpHeaderParser.parseRangeHeader to parse copy source range.

bcully committed:
Make max copy size before multipart a setting.

And override it in S3RepositoryAnalysisRestIT to
exercise multipart copy.

bcully committed:
Merge commit '3bac50e' into ES-11059

DaveCTurner reviewed (approved):
LGTM ~except, please can we backport this to 8.19 too? It's going to make the AWS SDK upgrade even more painful than it already is if we have this extra API in 9.1 vs in 8.19.~ scratch that, we discussed elsewhere - we'll work around the conflicts somehow

bcully committed:
Fix blobstore mock in S3BlobStoreContainerTests::testCopy
